### ISSUE TITLE
invalid-key when iterating keys of a TypedDict and indexing into the TypedDict

### ISSUE THREAD

**Issue by brent-at-aam** [astral-sh/ty#3378]
### Summary

Apologies if this has been covered. I did search for invalid-key errors and none seemed applicable. Consider a TypedDict defined as:

```python
class Foo(TypedDict):
  a: int
  b: str
```

If I iterate over the keys of an instance of Foo, I get an invalid-key error since the key is just a `str` and not a string literal.

```python
foo = Foo(a=1, b="hello world")
for k in foo:
  print(foo[k]) # invalid-key str not literal string
```

It seems that this is something that should be allowed given that k is `KeyOf<Foo>`. 

### Version

0.0.33

**Comment by ibraheemdev** [comment]
Hi, thanks for the report. For typed dicts that are `closed=True`, we could infer a literal type for `TypedDict.keys()/items()`, which would make your example work. That is tracked in https://github.com/astral-sh/ty/issues/2703. For open typed dicts, it seems possible to special case this specific pattern by checking if the place `k` originated from the same typed dict being subscripted, but that seems quite fragile, and something I don't think any other type checker does.

**Comment by sharkdp** [comment]
I'm going to close this as a duplicate of #2703, since the discussion there seems to be about both open and closed TypedDicts. Ibraheem, I will copy your suggestion into that ticket.
